hips/hip-9999.md

```diff
@@ -0,0 +1,102 @@
+---
+hip: 9999
+title: "Autorecover from stuck situations"
+authors: [ "Gernot Feichter <gernotfeichter@gmail.com>" ]
+created: "2024-07-12"
+type: "feature"
+status: "draft"
+helm-version: 3
+---
+
+## Abstract
+
+The idea is to simplify the handling for both manual users and CI/CD pipelines,
+to auto-recover from a state of stuck deployments, which is currently not possible unless users implement
+boilerplate code around their helm invocations.
+
+## Motivation
+
+If a helm deployment fails, I want to be able to retry it,
+ideally by running the same command again to keep things simple.
+
+There are two known situations how the user can run into such a situation where a retry will NOT work:
+1. A helm upgrade/install process is killed while the release is in state `PENDING-UPGRADE` or `PENDING-INSTALL`.
+2. The initial helm release installation (as performed via `helm upgrade --install`) is in state `FAILED`.
+
+Known Workarounds that should become OBSOLETE when this HIP is implemented to recover from such situations:
+1. `kubectl delete secret '<the name of the secret where helm stores release information>'.` (Not possible if you don't want to lose all history)
+2. `helm delete` your release. (Not possible if you don't want to lose all history)
+3. `helm rollback` your release. (Not possibly if it is the first installation)
+
+## Rationale
+
+The proposed solution uses a locking mechanism that is stored in k8s, such that all clients know whether a helm
+release is locked by themselves or not and for how long the lock is valid.
+
+It uses existing helm parameters like the --timeout parameter (which defaults to 5m) to determine for how long a helm release
+may be stuck in a pending state.
+
+## Specification
+
+The --timout parameter gets a deeper meaning.
+Previously the --timout parameter only had an effect on the helm process running on the respective client.
+After implementation, the --timout parameter will be stored in the helm release object (secret) in k8s and
+have an indirect impact on possible parallel processes.
+
+`helm ls -a` shows two new columns, regular `helm ls` does NOT show those:
```

Comment:
I believe that would not be fully backwards compatible. Other programs that depend on Helm may parse the output from `helm ls -a` and expect the number of columns to be unchanged.

Not sure how bad that is, though. [Every change is a breaking change to someone](https://xkcd.com/1172/), and I think this seems like a reasonable change.  And parsing the default (table) output is not robust anyway, if they do parse the output they should probably use `-o json` or `-o yaml` instead.

Reply:
I am very open to change the behaviour regarding `-a`, since `-a` traditionally means all releases (in all states), rather than the colums displayed.

Maybe rather tie it to the existing flag `--debug` or add a new flag? I actually would prefer that, maybe call it `--concurrency`? It's really just that most users will never need to see it...

@lindhe :  What do you think?

Sidenote: There is also the `--short` flag, but that false per default, so we cannot use that and having a long flag could yield to funny situations, e.g. specifying short and long....

Reply:
I'm impartial to changing it or not. If this feature can get added with the modification to `-a`, I think that's a good change. But if that causes any roadblocks I do not think it's worth it.

I think adding a new flag sounds a bit convoluted. If it cannot be part of `-a` by default , then it might make sense to tie into `--debug` like you suggest. But I would like to get some input from Helm maintainers before going too far with the specifics of that implementation. If they agree that it is OK to introduce a new column by default, that's probably the cleanest solution.